### PR TITLE
Compare arg+aliases between docs and argument_spec

### PR DESCRIPTION
**Pull request by sivel**
##### SUMMARY

This PR adds functionality to compare arg+aliases between `DOCUMENTATION.options` and the `argument_spec` in modules.

It should account for `FILE_COMMON_ARGUMENTS` as well as documentation fragments.  It's possible that something else was missed.

Of note here, is that there is no differentiation between the parent arg, and the aliases, they are treated the same.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
test/sanity/validate-modules

##### ANSIBLE VERSION

```
devel
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by gundalow** [comment]
1) 
`ERROR: lib/ansible/modules/network/nxos/nxos_nxapi.py:0:0: E323 "use_proxy" is listed in DOCUMENTATION.options, but not accepted by the module`

however
`'use_proxy': dict(default=True, type='bool'),` is in lib/ansible/module_utils/network/nxos/nxos.py

**Comment by gundalow** [comment]
2) 
network/f5 seems to be confused
`ERROR: lib/ansible/modules/network/f5/bigip_irule.py:0:0: E322 "pass" is listed in the argument_spec, but not documented in the module`

I don't see where that is defined

**Comment by sivel** [comment]
> ERROR: lib/ansible/modules/network/nxos/nxos_nxapi.py:0:0: E323 "use_proxy" is listed in DOCUMENTATION.options, but not accepted by the module

This is due to it being documented as a top level argument, but not accepted as a top level argument. It's only accepted within `provider`



> ERROR: lib/ansible/modules/network/f5/bigip_irule.py:0:0: E322 "pass" is listed in the argument_spec, but not documented in the module
> 
> I don't see where that is defined

It is defined in https://github.com/ansible/ansible/blob/2416a8806dbe28460fb860137e07e6e1d60166c0/lib/ansible/module_utils/f5_utils.py#L44-L50 and get's brought along as part of the `AnsibleF5Client` instantiation

**Comment by gundalow** [comment]
Though for 2 we have
https://github.com/ansible/ansible/blob/devel/lib/ansible/utils/module_docs_fragments/f5.py#L23 which defines `password:` in the docs. 
Seems like we are incorrectly checking that aliases exist as first-class options in `DOCUMENTATION`

**Comment by sivel** [comment]
> Though for 2 we have
> https://github.com/ansible/ansible/blob/devel/lib/ansible/utils/module_docs_fragments/f5.py#L23 which defines password: in the docs.
> Seems like we are incorrectly checking that aliases exist as first-class options in DOCUMENTATION

From the PR description:

> Of note here, is that there is no differentiation between the parent arg, and the aliases, they are treated the same.

This can be modified, but currently, it is ensuring that things match, and there is no documentation of those aliases.  If the documentation contained the `pass` and `pwd` aliases in DOCUMENTATION, the error wouldn't occur, as it is, the module accepts argument names, that aren't documented.

**Comment by gundalow** [comment]
Oh, I see, it's saying we are missing `aliases: [pass]` from https://github.com/ansible/ansible/blob/e6a7c0e30be95c47d9385b4ee330d842335a4fd0/lib/ansible/utils/module_docs_fragments/f5.py#L23-L28

**Comment by sivel** [comment]
> Oh, I see, it's saying we are missing aliases: [pass] from

bingo!

**Comment by gundalow** [comment]
I've started to fix the issues this has found in https://github.com/ansible/ansible/compare/devel...gundalow:docs-fix-args?expand=1

5543 of the 7168 issues are with `network/aci`, which is already known about and will be fixed by module freeze (if not before)

**Comment by sivel** [comment]
Tagged as a WIP, while we evaluate how this is working.

**Comment by gundalow** [comment]
Part one of fixes identified by this https://github.com/ansible/ansible/pull/34895
How about:

1. Review and merge https://github.com/ansible/ansible/pull/34895 (fixes part 1
2. Review and merge this PR (updating ignore list)
3. @gundalow will fix `provider` for network modules
4. We work though the remaining issues

With the aim of getting most of these documentation issues fixed before we branch `stable-2.5`

**Comment by gundalow** [comment]
https://github.com/ansible/ansible/pull/34895 Part 1 of fixes merged (general)
https://github.com/ansible/ansible/pull/34911 Part 2 of fixes merged (network)

So I think it's now time to rebase and merge this PR. Thanks again for the extra tests :)

File: test/sanity/validate-modules/main.py [hunk 1]
```diff
@@ -1020,6 +1032,50 @@ def _validate_argument_spec(self):
                          'should not be marked as required' % arg)
                 )
 
+        if docs:
+            try:
+                add_fragments(docs, self.object_path)
+            except Exception:
+                # Cannot merge fragments
```
Comment (gundalow) [inline test/sanity/validate-modules/main.py]: The code does seem to work for `extends_documentation_fragment`.

What does fragments mean in this context, can the comment be fleshed out a bit?

**Review by gundalow — COMMENTED** [review]
Just one comment

File: test/sanity/validate-modules/main.py (hunk 1 above)
Comment (sivel) [reply test/sanity/validate-modules/main.py]: This should work, unless it was as a result of the "temporary" PR being merged, which changed the signature of `add_fragments` slightly, that could have caused this to not work.

I've compared the docs of a module before and after this call, and confirmed that the fragments are properly merged

**Comment by sivel** [comment]
@gundalow I have this rebased and up to date as of now, if you are ready, give me the thumbs up, or go ahead and merge.

**Review by gundalow — APPROVED** [review]

File: test/sanity/validate-modules/main.py [hunk 2]
```diff
@@ -1011,7 +1013,17 @@ def _validate_argument_spec(self):
             )
             return
 
+        provider_args = set()
+        args_from_argspec = set()
         for arg, data in spec.items():
+            args_from_argspec.add(arg)
+            args_from_argspec.update(data.get('aliases', []))
+            if arg == 'provider':
```
Comment (mattclay) [inline test/sanity/validate-modules/main.py]: Shouldn't this be constrained to network modules, to avoid a false hit on the arg in non-network modules?

**Review by mattclay — CHANGES_REQUESTED** [review]

File: test/sanity/validate-modules/main.py (hunk 2 above)
Comment (sivel) [reply test/sanity/validate-modules/main.py]: Do you have a specific recommendation on detecting that?

Comment (mattclay) [reply test/sanity/validate-modules/main.py]: Checking to see if the module resides under the `lib/ansible/modules/network/` directory should work.

Comment (sivel) [reply test/sanity/validate-modules/main.py]: This has been addressed.

**Review by mattclay — APPROVED** [review]

**Pull request merged**